src/typechecker.erl

```diff
@@ -1385,7 +1397,7 @@ type_check_block_in(Env, ResTy, [Expr]) ->
 type_check_block_in(Env, ResTy, [Expr | Exprs]) ->
     {_, VarBinds, Cs1} = type_check_expr(Env, Expr),
     {VB, Cs2} = type_check_block_in(Env#env{ venv = add_var_binds(Env#env.venv, VarBinds) }, ResTy, Exprs),
-    {VB, constraints:combine(Cs1, Cs2)}.
+    {union_var_binds([Env#env.venv, VarBinds, VB]), constraints:combine(Cs1, Cs2)}.
```

Comment:
I'm not sure which variable bindings `type_check_expr[_in]` and all similar functions should return
- only new ones bound in the expression checked
- a cumulated set that also includes variable bindings from the input `Env`

Reply:
It should only return new one bound in the expression being checked. That means, `Env#env.venv` should not be in there.

You've clearly identified a bug here! The `VarBinds` clearly needs to be returned as well, just as you have written. Good job!